vent/controller/control_module.py

```
@@ -319,148 +234,98 @@ def __analyze_last_waveform(self):
+                self._DATA_BPM = 60. / phase[-1]  # 60 sec divided by the duration of last waveform, exception if this was 0.
+            except:
+                self._DATA_BPM = np.nan
 
     def get_sensors(self) -> SensorValues:
         # Make sure to return a copy of the instance
-        self._lock.acquire()
-        #cp = copy.deepcopy( self.COPY_sensor_values )
-        # don't need to deepcopy because a new SensorValues object is created
-        # each time and it copies each individual value as it's created
-        cp = copy.copy(self.COPY_sensor_values)
-        self._lock.release()
+        with self._lock:
+            cp = copy.copy(self.COPY_sensor_values)
+        self._time_last_contact = time.time()
         return cp
 
-    # def get_alarms(self) -> List[Alarm]:
-    #     # Returns all alarms as a list
-    #     self._lock.acquire()
-    #     new_alarm_list = self.COPY_logged_alarms.copy()
-    #     for alarm_key in self.COPY_active_alarms.keys():
-    #         new_alarm_list.append(self.COPY_active_alarms[alarm_key])
-    #     self._lock.release()
-    #     return new_alarm_list
-
-    def get_active_alarms(self):
-        # Returns only the active alarms
-        self._lock.acquire()
-        active_alarms = self.COPY_active_alarms.copy() # Make sure to return a copy
-        self._lock.release()
-        return active_alarms
-
-    def get_logged_alarms(self) -> List[Alarm]:
-        # Returns only the inactive alarms
-        self._lock.acquire()
-        logged_alarms = self.COPY_logged_alarms.copy()
-        self._lock.release()
-        return logged_alarms
-
+    def get_alarms(self):
+        """
+        Returns alarms, by time of occurance:
+        """
+        with self._lock:
+            hapa = self.HAPA
+            techa = self.TECHA
 
+        if (hapa is not None) or (techa is not NONE):
+            print("An alarm has happened")
+            #TODO: Link to Alarm manager
+        else:
+            return None
```

@sneakers-the-rat Something like this? How should the alarms be raised?

---

controller shouldn't do anything -- just let it be raised by alarm manager